query-engine/core/src/query_graph_builder/write/upsert.rs

```diff
@@ -200,66 +200,53 @@ pub fn upsert_record(
 // 4. The unique field defined in where clause has the same value as defined in the create arguments
 fn can_use_connector_native_upsert(
     model: &ModelRef,
-    where_field: ParsedInputMap,
-    create_argument: ParsedInputMap,
-    update_argument: ParsedInputMap,
+    where_field: &ParsedInputMap,
+    create_argument: &ParsedInputMap,
+    update_argument: &ParsedInputMap,
     connector_ctx: &ConnectorContext,
 ) -> bool {
     let has_nested_create = create_argument
-        .clone()
-        .into_iter()
-        .any(|(k, _)| matches!(model.fields().find_from_all(&k), Ok(Field::Relation(_))));
+        .iter()
+        .any(|(field_name, _)| model.fields().find_from_relation_fields(&field_name).is_ok());
 
     let has_nested_update = update_argument
-        .into_iter()
-        .any(|(k, _)| matches!(model.fields().find_from_all(&k), Ok(Field::Relation(_))));
+        .iter()
+        .any(|(field_name, _)| model.fields().find_from_relation_fields(&field_name).is_ok());
 
-    let uniques = where_field
-        .clone()
-        .into_iter()
+    let has_one_unique = where_field
+        .iter()
         .filter(|(field_name, input)| {
             is_unique_field(field_name, model) && where_and_create_equal(field_name, input, &create_argument)
         })
-        .count();
-
-    let where_values_same_as_create = where_field
-        .clone()
-        .into_iter()
-        .all(|(field_name, input)| where_and_create_equal(&field_name, &input, &create_argument));
+        .count()
+        == 1;
```

Comment:
Doesn't your refactor change your logic? This currently ensures that there is _only one where value that's equal to a create value_.

So, given this query, it would still pass:
```graphql
upsertOneUser(where: { unique1: 1, unique2: 2 }, create: { unique1: 1, unique2: 3 })
```
Although, IIUC, it shouldn't.


IIUC, what you want to check is:
1. That there is _only_ 1 unique field in the where (whether their value is equal to the create one shouldn't be taken into account).
2. That this single unique field is the same in the create

Reply:
Good catch